python/ray/ml/train/integrations/huggingface/huggingface_torch_trainer.py

```diff
@@ -0,0 +1,217 @@
+
+            def tokenize_function(examples):
+                return tokenizer(examples["text"])
+
+            tokenized_datasets = datasets.map(
+                tokenize_function, batched=True, num_proc=1, remove_columns=["text"]
+            )
+
+            def group_texts(examples):
+                # Concatenate all texts.
+                concatenated_examples = {
+                    k: sum(examples[k], []) for k in examples.keys()
+                }
+                total_length = len(concatenated_examples[list(examples.keys())[0]])
+                # We drop the small remainder, we could add padding if the model
+                # supported it.
+                # instead of this drop, you can customize this part to your needs.
+                total_length = (total_length // block_size) * block_size
+                # Split by chunks of max_len.
+                result = {
+                    k: [
+                        t[i : i + block_size]
+                        for i in range(0, total_length, block_size)
+                    ]
+                    for k, t in concatenated_examples.items()
+                }
+                result["labels"] = result["input_ids"].copy()
+                return result
+
+            lm_datasets = tokenized_datasets.map(
+                group_texts,
+                batched=True,
+                batch_size=1000,
+                num_proc=1,
+            )
+            ray_train_ds = ray.data.from_arrow(lm_datasets["train"]._data.table)
+            ray_validation_ds = ray.data.from_arrow(
+                lm_datasets["validation"]._data.table
+            )
+
+            def pre_init_function(args, trainer_kwargs):
+                model_config = AutoConfig.from_pretrained(model_checkpoint)
+                model = AutoModelForCausalLM.from_config(model_config)
+                trainer_kwargs["model"] = model
+
+            scaling_config = {"num_workers": 3}
+            # If using GPUs, use the below scaling config instead.
+            # scaling_config = {"num_workers": 3, "use_gpu": True}
+            trainer = HuggingFaceTorchTrainer(
+                args={
+                    "output_dir": f"{model_checkpoint}-wikitext2",
+                    "evaluation_strategy": "epoch",
+                    "learning_rate": 2e-5,
+                    "weight_decay": 0.01,
+                },
+                pre_init_function=pre_init_function,
+                scaling_config=scaling_config,
+                datasets={"train": ray_train_ds, "validation": ray_validation_ds},
+            )
+            result = trainer.fit()
+
+    Args:
+        trainer_class: A subclass of ``transformers.Trainer`` to use.
+            Defaults to ``transformers.Trainer``.
+        args: Keyword arguments passed to ``transformers.TrainingArguments``
+            during initialization on every Actor.
+        pre_init_function: A function taking in the ``args`` and ``trainer_kwargs``
```

Comment:
@Yard1 one alternatve is to just accept a `trainer_init_fn` that returns an instatiated transformers.Trainer. Then we don't have to accept `pre_init_function`, `trainer_class`, or `trainer_kwargs`- these would all be encapsulated in the function.

Reply:
Yep passing a factory should avoid the problem entirely.